src/Doctrine/Odm/Tests/Extension/PaginationExtensionTest.php

```diff
@@ -410,16 +431,17 @@ private function mockAggregationBuilder(int $expectedOffset, int $expectedLimit)
         $skipProphecy = $this->prophesize(Skip::class);
         if ($expectedLimit > 0) {
             $skipProphecy->limit($expectedLimit)->shouldBeCalled();
-        } else {
-            $matchProphecy = $this->prophesize(AggregationMatch::class);
-            $matchProphecy->field(Paginator::LIMIT_ZERO_MARKER_FIELD)->shouldBeCalled()->willReturn($matchProphecy->reveal());
-            $matchProphecy->equals(Paginator::LIMIT_ZERO_MARKER)->shouldBeCalled()->willReturn($matchProphecy->reveal());
-            $skipProphecy->match()->shouldBeCalled()->willReturn($matchProphecy->reveal());
         }
 
         $resultsAggregationBuilderProphecy = $this->prophesize(Builder::class);
         $resultsAggregationBuilderProphecy->skip($expectedOffset)->shouldBeCalled()->willReturn($skipProphecy->reveal());
 
+        $addFieldsProphecy = $this->prophesize(AddFields::class);
```

Comment:
@alcaeus The class [`Doctrine\ODM\MongoDB\Aggregation\Stage\AddFields`](https://github.com/doctrine/mongodb-odm/blob/2.10.x/lib/Doctrine/ODM/MongoDB/Aggregation/Stage/AddFields.php) is final, it's impossible to mock. I had to duplicate it here.

I hate how this tests are mocking the full aggregation builder. It's very hard to read and to update. It should be refactored to use an actual MongoDB connection.

Reply:
I removed the duplicated class thanks to https://github.com/doctrine/mongodb-odm/pull/2717.